File: src/main/java/de/tum/cit/aet/artemis/atlas/dto/CourseCompetencyDTO.java
package de.tum.cit.aet.artemis.atlas.dto;

import jakarta.annotation.Nullable;
import jakarta.validation.constraints.NotNull;

import com.fasterxml.jackson.annotation.JsonInclude;

import de.tum.cit.aet.artemis.atlas.domain.competency.CompetencyTaxonomy;
import de.tum.cit.aet.artemis.atlas.domain.competency.CourseCompetency;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
public record CourseCompetencyDTO(long id, @NotNull String title, @Nullable String description, @Nullable CompetencyTaxonomy taxonomy) {

    public static CourseCompetencyDTO of(CourseCompetency competency) {
        return new CourseCompetencyDTO(competency.getId(), competency.getTitle(), competency.getDescription(), competency.getTaxonomy());
    }
}

File: src/main/webapp/app/atlas/shared/dto/course-competency-dto.ts
import { CompetencyTaxonomy, CourseCompetency } from 'app/atlas/shared/entities/competency.model';

export interface CourseCompetencyDTO {
    id: number;
    title: string;
    description?: string;
    taxonomy?: CompetencyTaxonomy;
}

/**
 * Convert CourseCompetency (client model) → CourseCompetencyDTO (API DTO).
 */
export function toCourseCompetencyDTO(competency: CourseCompetency): CourseCompetencyDTO {
    if (competency.id === undefined) {
        throw new Error('Cannot map CourseCompetency to DTO: id is missing.');
    }
    if (!competency.title) {
        throw new Error('Cannot map CourseCompetency to DTO: title is missing.');
    }
    return {
        id: competency.id,
        title: competency.title,
        description: competency.description ?? undefined,
        taxonomy: competency.taxonomy ?? undefined,
    };
}
